staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go

```diff
@@ -33,28 +33,22 @@ func NewStatusStrategy(strategy customResourceStrategy) statusStrategy {
 }
 
 func (a statusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	// update is only allowed to set status
 	newCustomResourceObject := obj.(*unstructured.Unstructured)
-	oldCustomResourceObject := old.(*unstructured.Unstructured)
-
 	newCustomResource := newCustomResourceObject.UnstructuredContent()
-	oldCustomResource := oldCustomResourceObject.UnstructuredContent()
-
-	// update is not allowed to set spec and metadata
-	_, ok1 := newCustomResource["spec"]
-	_, ok2 := oldCustomResource["spec"]
-	switch {
-	case ok2:
-		newCustomResource["spec"] = oldCustomResource["spec"]
-	case ok1:
-		delete(newCustomResource, "spec")
-	}
+	status, ok := newCustomResource["status"]
 
-	newCustomResourceObject.SetAnnotations(oldCustomResourceObject.GetAnnotations())
-	newCustomResourceObject.SetFinalizers(oldCustomResourceObject.GetFinalizers())
-	newCustomResourceObject.SetGeneration(oldCustomResourceObject.GetGeneration())
-	newCustomResourceObject.SetLabels(oldCustomResourceObject.GetLabels())
-	newCustomResourceObject.SetOwnerReferences(oldCustomResourceObject.GetOwnerReferences())
-	newCustomResourceObject.SetSelfLink(oldCustomResourceObject.GetSelfLink())
+	// copy old object into new object
+	oldCustomResourceObject := old.(*unstructured.Unstructured)
+	oldCustomResourceObject.DeepCopyInto(newCustomResourceObject)
```

Comment:
Doesn't this override the `ResourceVersion` in metadata, i.e. break the optimistic concurrency?

Reply:
We have checked the resourceVersion [here](https://github.com/kubernetes/kubernetes/blob/37ef3232b9390ac025797f040ee33a73546e7708/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L610), so I think newCustomResourceObject and oldCustomResourceObject should have the same resourceVersion here in this function.

Reply:
Can you add a small comment that the resource version is correct here?

Reply:
Just noticed that this `DeepCopyInto` is potentially dangerous. By definition, `DeepCopyInto` might reuse the target object. So this call might override the status that we saved 3 lines above (right now, the `DeepCopyInto` implementation for `Unstructured` is not optimizied. So this works. But that's a conincidence).

Better use `newCustomResourceObject = *oldCustomResourceObject.DeepCopy()` (it's what happens today in `Unstructured.DeepCopyInto`).

Reply:
Done.

Reply:
Thanks very much for you detailed guidence!!!